### PR TITLE
CA-201728: VLAN.create must fail if VLAN is in use by FCoE

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

**Comment by sharady** [comment]
This PR depends on https://github.com/xapi-project/xcp-idl/pull/122 and https://github.com/xapi-project/xcp-networkd/pull/86

File: ocaml/xapi/xapi_vlan.ml [hunk 1]
```diff
@@ -63,6 +63,11 @@ let create ~__context ~tagged_PIF ~tag ~network =
 	if Db.PIF.get_tunnel_access_PIF_of ~__context ~self:tagged_PIF <> [] then
 		raise (Api_errors.Server_error (Api_errors.is_tunnel_access_pif, [Ref.string_of tagged_PIF]));
 
+	(* Check the VLAN is not in use by FCoE *)
+	let vlan_str = device ^ "." ^ (Int64.to_string tag) in
```
Comment (lindig) [inline ocaml/xapi/xapi_vlan.ml]: Consider using `Printf.sprintf "%s.%Ld" device tag`

**Comment by lindig** [comment]
This looks solid.

File: ocaml/xapi/xapi_vlan.ml [hunk 2]
```diff
@@ -63,6 +63,11 @@ let create ~__context ~tagged_PIF ~tag ~network =
 	if Db.PIF.get_tunnel_access_PIF_of ~__context ~self:tagged_PIF <> [] then
 		raise (Api_errors.Server_error (Api_errors.is_tunnel_access_pif, [Ref.string_of tagged_PIF]));
 
+	(* Check the VLAN is not in use by FCoE *)
```
Comment (robhoes) [inline ocaml/xapi/xapi_vlan.ml]: Make this comment more general: "...in use by the kernel".

**Comment by robhoes** [comment]
You also need to handle the case of a pre-existing VLAN in the DB, where `PIF.plug` is fails if the VLAN is in use by the kernel, e.g. when someone first creates the VLAN on XS, and then reboots with FCoE enabled. This will end up as a failed call to `Bridge.make_config` in `Nm.bring_pif_up`. I suggest adding a new exception to network_interface.ml, and getting xcp-networkd to raise that exception in this situation.

File: ocaml/xapi/xapi_vlan.ml [hunk 3]
```diff
@@ -13,6 +13,7 @@
  *)
 module D = Debug.Make(struct let name="xapi" end) 
 open D
+open Network
```
Comment (lindig) [inline ocaml/xapi/xapi_vlan.ml]: I'd prefer a local `let open Network in` where you need it to keep the code readable without tool support.

**Comment by robhoes** [comment]
You are still not handling the `PIF.plug` case.

**Comment by sharady** [comment]
@robhoes @lindig thanks for the review, I have updated the PR.

**Comment by lindig** [comment]
I'm adding the `reviewed` flag but would like to ask @robhoes to also take another look.

File: ocaml/xapi/network.ml [hunk 4]
```diff
@@ -42,6 +42,9 @@ let transform_networkd_exn pif f =
 	| Not_implemented ->
 		let e = "networkd function not implemented" in
 		reraise Api_errors.pif_configuration_error [Ref.string_of pif; e]
+	| Vlan_in_use params ->
+		let e = String.concat "; " (List.map (fun (k, v) -> k ^ " = " ^ v) params) in
+		reraise Api_errors.vlan_in_use [Ref.string_of pif; e]
```
Comment (robhoes) [inline ocaml/xapi/network.ml]: I don't think the PIF ref is appropriate here. The datamodel says that the first param is the device. Also, if you change the arguments of the networkd exception to `(string * int)` (see xapi-project/xcp-idl#122), then you can just do:
```
| Vlan_in_use (device, vlan) ->
    reraise Api_errors.vlan_in_use [device; vlan]
```

**Pull request merged**